### PR TITLE
Refactor message converter config

### PR DESCRIPTION
**Pull request by ilayaperumalg**
- Move message converter configuration into a separate class `MessageConverterConfigurer`

### CODE REVIEW

File: spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binding/ChannelBindingService.java [hunk 1]
```diff
@@ -53,33 +32,16 @@
...
 
+	@Autowired
 	private Binder<MessageChannel> binder;
 
+	@Autowired
 	private ChannelBindingServiceProperties channelBindingServiceProperties;
 
-	private CompositeMessageConverterFactory messageConverterFactory;
-
-	public ChannelBindingService(ChannelBindingServiceProperties channelBindingServiceProperties, Binder<MessageChannel> binder) {
```
Comment (mbogoevici) [inline spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binding/ChannelBindingService.java]: Kind of a nit, but I think we should keep the `Binder` and `ChannelBindingServiceProperties` configured via constructor   to keep things clear (for that matter we could make them final too). The mix of no-arg construction in `ChannelBindingServiceConfiguration` and field autowiring makes this a little hard to follow, IMO.

File: spring-cloud-stream/src/main/java/org/springframework/cloud/stream/config/ChannelBindingServiceConfiguration.java [hunk 2]
```diff
@@ -64,10 +65,10 @@
 	// it is used to detect a ChannelBindingService in the parent context (which we know
 	// already exists).
 	@ConditionalOnMissingBean(ChannelBindingService.class)
-	public ChannelBindingService bindingService(
-			ChannelBindingServiceProperties channelBindingServiceProperties,
+	public ChannelBindingService bindingService(ChannelBindingServiceProperties channelBindingServiceProperties,
 			Binder<MessageChannel> binder) {
-		return new ChannelBindingService(channelBindingServiceProperties, binder);
+		return new ChannelBindingService(channelBindingServiceProperties, binder,
+				messageConvertersConfigurer(channelBindingServiceProperties));
```
Comment (mbogoevici) [inline spring-cloud-stream/src/main/java/org/springframework/cloud/stream/config/ChannelBindingServiceConfiguration.java]: I think we can rather pass the instance as an argument to the `bindingService()` method instead.

**Comment by mbogoevici** [comment]
Looking at this, I have the feeling that we can take this a step further and extract the entire channel creation logic from `BindableProxyFactory` in a `ChannelFactory`, that the  `BindableProxyFactory`  could delegate to for creating channels (including adding converters) - then we wouldn't need the separate `MessageConverterConfigurer` and repeated calls to `ChannelBindingService`. (In theory, we could handle channel creation as part of `ChannelBindingService`, but it would be nice to have those responsibilities separated.

**Comment by ilayaperumalg** [comment]
That's a good idea. But I believe it is not straightforward to have `ChannelBindingServiceProperties` in proxy factory class to do message converter configuration.

**Comment by mbogoevici** [comment]
@ilayaperumalg that wouldn't be necessary. `ChannelBindingServiceProperties` would be injected in the `ChannelFactory`

**Comment by ilayaperumalg** [comment]
Closing this PR to issue a new one that uses `ChannelFactory`.
